alf/algorithms/mcts_models.py

```diff
@@ -174,8 +192,18 @@ def recurrent_inference(self, state, action):
         Returns:
             ModelOutput: the prediction
         """
-        new_state = self._dynamics_net((state, action))[0]
-        return self.prediction_model(new_state)
+        if not self._handle_bn:
+            new_state = self._dynamics_net((state, action))[0]
+            return self.prediction_model(new_state)
+        else:
+            state, current_steps = state
+            self._prediction_net.set_batch_norm_current_step(current_steps)
+            new_state = self._dynamics_net((state, action))[0]
+            current_steps = current_steps + 1
+            self._dynamics_net.set_batch_norm_current_step(current_steps)
```

Comment:
Is the first step BN of ``_dynamics_net``  is not used? And the effective number of BNs is ``num_unroll_steps - 1`` for ``_dynamics_net``?

Reply:
dynamics_net will be called  `num_unroll_steps` times. prediction_net will be called `num_unroll_steps+1` times.

Reply:
Yes, I understand that the dynamics_net will be called ``num_unroll_steps`` times and therefore we have set
``self._dynamics_net.set_batch_norm_max_steps(num_unroll_steps)``.

What I meant is that there seems to be an index offset.

It seems to be that there will be a clamp issue in the end of recurrent inference.
If the recurrent inference has a total length of ``num_unroll_steps``,  then in the last step, 
when we call ``self._dynamics_net.set_batch_norm_current_step(current_steps)``, the  ``current_steps`` now has the value of ``num_unroll_steps``, which will be clamped to ``num_unroll_steps - 1`` in ``set_batch_norm_max_steps``.

On the other hand, the 0-th BN statistics for ``dynamics_net`` does not seems to be involved here (as ``current_steps`` starts from 1 because of Line213).

Reply:
For the first call of  `self._dynamics_net.set_batch_norm_current_step(current_steps)`, `current_step` is 0. For the last call, `current_step` is `num_unroll_steps - 1`